internal/cmd/controller/reconciler/bundle_controller.go

```diff
@@ -496,26 +505,43 @@ func (r *BundleReconciler) updateStatus(ctx context.Context, orig *fleet.Bundle,
 	return nil
 }
 
-// cleanupOrphanedBundleDeployments will delete all existing BundleDeployments which do not have a match in a provided list of UIDs
-func (r *BundleReconciler) cleanupOrphanedBundleDeployments(ctx context.Context, bundle *fleet.Bundle, uidsToKeep sets.Set[types.UID]) error {
+func (r *BundleReconciler) listBundleDeploymentsForBundle(ctx context.Context, bundle *fleet.Bundle) ([]fleet.BundleDeployment, error) {
 	list := &fleet.BundleDeploymentList{}
 	if err := r.List(ctx, list,
 		client.MatchingLabels{
 			fleet.BundleLabel:          bundle.GetName(),
 			fleet.BundleNamespaceLabel: bundle.GetNamespace(),
 		},
 	); err != nil {
+		return nil, err
+	}
+	return list.Items, nil
+}
+
+// cleanupOrphanedBundleDeployments will delete all existing BundleDeployments which do not have a match in a provided list of UIDs
+func (r *BundleReconciler) cleanupOrphanedBundleDeployments(ctx context.Context, bundle *fleet.Bundle, uidsToKeep sets.Set[types.UID]) error {
+	list, err := r.listBundleDeploymentsForBundle(ctx, bundle)
+	if err != nil {
 		return err
 	}
+	toDelete := slices.DeleteFunc(list, func(bd fleet.BundleDeployment) bool {
+		return uidsToKeep.Has(bd.UID)
+	})
+	return batchDeleteBundleDeployments(ctx, r.Client, toDelete)
+}
 
+func batchDeleteBundleDeployments(ctx context.Context, c client.Client, list []fleet.BundleDeployment) error {
 	var errs []error
-	for _, bd := range list.Items {
-		if uidsToKeep.Has(bd.UID) {
+	for _, bd := range list {
+		if bd.DeletionTimestamp != nil {
```

Comment:
nit: why check the `DeletionTimestamp` against `nil` here instead of using `IsZero()`?

Reply:
I think they are equivalent. `IsZero` uses a pointer receiver so it will work even if `DeletionTimestamp` is `nil`.
I guess it's a matter of style; personally I find it more idiomatic to check against nil, than a negated `IsZero`, slightly easier thought process 😅 

BTW, I searched in the `kubernetes/kubernetes` codebase, and found [78 occurrences for `!= nil`](https://github.com/search?q=repo%3Akubernetes%2Fkubernetes%20%22DeletionTimestamp%20!%3D%20nil%22&type=code) vs. [only 14 for `IsZero`](https://github.com/search?q=repo%3Akubernetes%2Fkubernetes+%22DeletionTimestamp.IsZero%22&type=code) 😄